### PR TITLE
Factor new notebook

### PR DESCRIPTION
**Pull request by markples**
This allows an investigation to start with just `#!import BenchmarkAnalysis.dib` and dev iteration on the reports to then add `#!import Reports.dib`.

This should be no-diff other than the factoring except for the following:
- Adding #!import commands
- Factor and reduce import/using lists
- Whitespace (lint errors) in the documentation md file

### CODE REVIEW

File: src/benchmarks/gc/GC.Infrastructure/Notebooks/Reports-Impl.dib [hunk 1]
```diff
@@ -0,0 +1,1756 @@
...
+    public static Metric<TSource> Promote<TOldSource>(Metric<TOldSource> metric, Func<TSource, IEnumerable<TOldSource>> oldExtract, Aggregation aggregation)
+        => new(extract: source => aggregation.Func(oldExtract(source).Select(metric.ExtractFunc).Where(NotNull).Select(value => value.Value)),
+            title: $"{aggregation.Title} of {metric.Title}",
+            unit: aggregation.UnitOverride ?? metric.Unit);
+}
+
+public static class Metrics
+{
+    public static Metric<IterationData> Promote(Metric<TraceGC> metric, Aggregation aggregation)
+        => Metric<IterationData>.Promote<TraceGC>(metric, iterationData => (IEnumerable<TraceGC>)iterationData.GCProcessData.GCs, aggregation);
```
Comment (markples) [inline src/benchmarks/gc/GC.Infrastructure/Notebooks/Reports-Impl.dib:482]: I don't know why the explicit `<TraceGC>` and `(IEnumerable<TraceGC>)` are required for the factored version to build.

File: src/benchmarks/gc/GC.Infrastructure/Notebooks/Reports-Impl.dib [hunk 2]
```diff
@@ -0,0 +1,1756 @@
...
+    }
+
+    public override string GetColorFamilyId(SeriesInfo<TraceGC> info, bool multipleMetrics) => $"_{info.Iteration}";
+    public override string GetSeriesTitle(SeriesInfo<TraceGC> info, string colorFamilyKey, bool multipleMetrics) => $"{colorFamilyKey}{GetColorFamilyId(info, multipleMetrics)}";
+
+    public override string GetChartTitle() => "Per-run behavior";
+
+    public override List<KeyValuePair<string, TraceGC>> GetDataSource(SeriesInfo<TraceGC> info,
+        Filter benchmarkFilter, IntFilter iterationFilter, ConfigIterationFilter configIterationFilter, Func<TraceGC, bool> dataFilter)
+        => ((IEnumerable<TraceGC>)info.IterationData.GCProcessData?.GCs).Where(gc => gc.GlobalHeapHistory != null).Where(dataFilter).Select(gc => KeyValuePair.Create("", gc));
```
Comment (markples) [inline src/benchmarks/gc/GC.Infrastructure/Notebooks/Reports-Impl.dib:888]: I don't know why the explicit `(IEnumerable<TraceGC>)` is required for the factored version to build.

File: src/benchmarks/gc/GC.Infrastructure/Notebooks/Reports-Impl.dib [hunk 3]
```diff
@@ -0,0 +1,1756 @@
...
+    public static Metric<TSource> Promote<TOldSource>(Metric<TOldSource> metric, Func<TSource, IEnumerable<TOldSource>> oldExtract, Aggregation aggregation)
+        => new(extract: source => aggregation.Func(oldExtract(source).Select(metric.ExtractFunc).Where(NotNull).Select(value => value.Value)),
+            title: $"{aggregation.Title} of {metric.Title}",
+            unit: aggregation.UnitOverride ?? metric.Unit);
+}
+
+public static class Metrics
+{
+    public static Metric<IterationData> Promote(Metric<TraceGC> metric, Aggregation aggregation)
+        => Metric<IterationData>.Promote(metric, iterationData => (IEnumerable<TraceGC>)iterationData.GCProcessData.GCs, aggregation);
```
Comment (markples) [inline src/benchmarks/gc/GC.Infrastructure/Notebooks/Reports-Impl.dib:482]: I don't know why the explicit `(IEnumerable<TraceGC>)` is required for the factored version to build.

File: src/benchmarks/gc/GC.Infrastructure/Notebooks/Reports-Impl.dib (hunk 2 above)
Comment (markples) [reply src/benchmarks/gc/GC.Infrastructure/Notebooks/Reports-Impl.dib:888]: fixed by keeping all imports/usings together

File: src/benchmarks/gc/GC.Infrastructure/Notebooks/Reports-Impl.dib (hunk 3 above)
Comment (markples) [reply src/benchmarks/gc/GC.Infrastructure/Notebooks/Reports-Impl.dib:482]: fixed by keeping all imports/usings together

**Review by mrsharm — APPROVED** [review]
LGTM!

**Pull request merged**